### PR TITLE
Fix default arguments for dynamically dispatched methods

### LINKED ISSUE #7880: default argument value for overridden methods
(This issue is a split-out version of a comment by @cassella from issue #7868).

@cassella says:

Besides visibility, scoping, and promotion, what about virtual functions with different default values for an argument?  For example,

```chapel
class C {
  proc foo(x = 10) { writeln("C.foo:", x); }
}

class D : C {
  proc foo(x = 20) { writeln("D.foo:", x); }
}

var d = new D();
d.foo();

var c: C = new D();
c.foo();
```

master currently prints

```
D.foo:20
D.foo:10
```

It's surprising me that the second call is being dynamically dispatched to ``D.foo()`` but picking up ``C.foo()``'s default value.

I think this is the behavior that the proposed change would require, since it'd have to evaluate the default value where the call is made, so the base class's method is the only one it knows about there.  But I'm wondering if that's really what should happen.

FWIW, the spec (16.2.5) says methods can be overridden and virtually dispatched if they have identical signatures.  But then it lists what's required to be considered identical: "names, types, and order".  It doesn't include default values.  (Or intents, for that matter.)  If the default value should be part of the identicalness, I'd think it either shouldn't compile, or shouldn't virtually dispatch in this case.

### LINKED ISSUE #15164: virtual dispatch and default param arguments
Continuing from #13962 , where we decided that the default value and virtual method should match. There are two related cases that do not work on master today.

Consider the following program:
```chpl
class Parent {
  proc method(arg:int = f()) {
    writeln("in Parent.method arg=", arg);
  }
}

class Child : Parent {
  override proc method(arg:int = g()) {
    writeln("in Child.method arg=", arg);
  }
}

proc main() {
  var x:Parent = new Child();
  x.method();
  //   run child's default (virtually dispatched) `g()` and so print out `in Child.method arg=1`

  var y = new Child();
  y.method(); // expect this to run `g` and so print `in Child.method arg=1`
}
```

Case 1:
If the functions invoked as default values return a param:
```chpl
proc f() param { return 0; }
proc g() param { return 1; }
```
Then the compiler incorrectly folds the calls away and the wrong default value is used for the ``x.method()`` call.

Case 2:
If the formals have the ``param`` intent, then ``x.method()`` has the wrong default value.

**Associated Future Test(s):**
[`test/functions/default-arguments/default-argument-class-override3a.chpl`](
           https://github.com/chapel-lang/chapel/blob/master/test/functions/default-arguments/default-argument-class-override3a.chpl
          ) #17653

### LINKED ISSUE #17649: Overridden method Named Arg defaults not being honored.
### Summary of Problem
This is an issue involving inheritance, overriding methods and named argument defaults values; a more complete description of the problem is below.  In short the default values in the overridden method are not being honored, the parent class's are still being used.  I'm not sure if this is expected behavior or a bug; if it's expected maybe the Named Arguments and Default Values documentation could be clarified (https://chapel-lang.org/docs/language/spec/procedures.html#named-arguments)

_**What behavior did you observe when encountering this issue?**_
I declared a base class Foo with a proc `hello` and an inheriting class Bar which overrides `hello`.  In proc `hello` I have named arguments with default values.  Foo has one set of defaults and Bar has another.  When I declare and create a Foo and test out the defaults I get Foo's defaults; when i declare Bar and construct a Bar and repeat the test I get Bar's defaults; so far so good, everything works as expected.  The issue I run into is when I declare a Foo but provide Bar as the concrete implementation.  When I invoke the `hello` proc, it runs the overridden `Bar.hello()` but it gets Foo's defaults.

_**What behavior did you expect to observe?**_
When declaring a Foo and providing Bar as the concrete impl. I expected to get Bar's default values for the named arguments since Bar's `hello` proc was the one that got invoked.

This may be working as the Chapel team intended but it was unexpected and the documentation on Named Arguments and Default Values (https://chapel-lang.org/docs/language/spec/procedures.html#named-arguments) didn't really mention anything about what happens in the case of inheritance.

_**Is this a blocking issue with no known work-arounds?**_
This is not a blocking issue, it's really easy to work around, it's more that it was unexpected behavior, see previous note about possibly intended behavior.

### Steps to Reproduce

**Source Code:**
Note, I also made a gist of it here: https://gist.github.com/glitch/ef9aa1f6643d4dd1429f8d7d4768906d

```chapel
/*
Chapel code example to illustrate issue with inheritance, overriding methods and named vars.
Declare a base class Foo with a hello proc.  It has 3 named args with defaults.
Next Declare Bar which inherits from Foo and overrides Foo's hello proc with its own
default values for the named args.
When we declare a Foo with a concrete implementation of Bar we see that the Bar hello proc
gets invoked as expected, but the default args do not; they are still Foo's defaults.
Since the overridden proc of Bar's gets invoked, I would have expected Bar's defaults to go
along with it, but that's not the case based on the output.
*/
class Foo {

    proc hello(greeting:string="foo default greeting", prefix:string="FOO_default_prefix", suffix:string="FOO_default_suffix"): string throws {
        return "FOO.hello(...)=> greeting:`" + greeting + "`, prefix:`" + prefix + "`,  suffix:`" + suffix + "`";
    }

}

class Bar : Foo {

    override proc hello(greeting:string="bar default greeting", prefix:string = "default_bar[", suffix:string = "]bar_default"): string throws {
        return "I'm in BAR.hello(...)=> greeting:`" + greeting + "`, prefix:`" + prefix + "`,  suffix:`" + suffix + "`";
    }
}


// Create Foo with concrete Foo impl
// 1. Prove we get the defaults
// 2. Pass in greeting and get defaults on prefix & suffix
// 3. Pass in greeting & prefix, get default on suffix
var f:Foo = new Foo();
writeln("----Foo----");
writeln(f.hello());
writeln(f.hello(greeting="hello"));
writeln(f.hello(greeting="hello", prefix="<<<"));

// Create Bar with concrete Bar impl
// 1. Prove we get Bar's defaults
// 2. Pass in greeting and get Bar's defaults on prefix & suffix
// 3. Pass in greeting & prefix, get Bar's default on suffix
var b:Bar = new Bar();
writeln("----Bar----");
writeln(b.hello());
writeln(b.hello(greeting="I'm Bar"));
writeln(b.hello(greeting="I'm Bar", prefix="<<<"));

// Create Foo with conreate Bar impl.
// This uses Bar's overridden proc hello, but doesn't get the Bar default named args, why?
// 1. We see it inokes Bar's hello method in the output, but we get Foo's named defaults
// 2. Pass in greeting, we see Bar's method invoked, but still get Foo defaults on prefix & suffix
// 3. Pass in greetin & suffix, same thing, We see Bar's output message, but still Foo default suffix.
var fb:Foo = new Bar();
writeln("----Foo with Bar as impl----");
writeln(fb.hello());
writeln(fb.hello(greeting="expect Bar defaults"));
writeln(fb.hello(greeting="expect Bar defaults minus passed in prefix", prefix="<<<"));
```

EDIT/UPDATE - Output of sample program:
```
----Foo----
FOO.hello(...)=> greeting:`foo default greeting`, prefix:`FOO_default_prefix`,  suffix:`FOO_default_suffix`
FOO.hello(...)=> greeting:`hello`, prefix:`FOO_default_prefix`,  suffix:`FOO_default_suffix`
FOO.hello(...)=> greeting:`hello`, prefix:`<<<`,  suffix:`FOO_default_suffix`
----Bar----
I'm in BAR.hello(...)=> greeting:`bar default greeting`, prefix:`default_bar[`,  suffix:`]bar_default`
I'm in BAR.hello(...)=> greeting:`I'm Bar`, prefix:`default_bar[`,  suffix:`]bar_default`
I'm in BAR.hello(...)=> greeting:`I'm Bar`, prefix:`<<<`,  suffix:`]bar_default`
----Foo with Bar as impl----
I'm in BAR.hello(...)=> greeting:`foo default greeting`, prefix:`FOO_default_prefix`,  suffix:`FOO_default_suffix`
I'm in BAR.hello(...)=> greeting:`expect Bar defaults`, prefix:`FOO_default_prefix`,  suffix:`FOO_default_suffix`
I'm in BAR.hello(...)=> greeting:`expect Bar defaults minus passed in prefix`, prefix:`<<<`,  suffix:`FOO_default_suffix`
```

**Compile command:**
`chpl default-args-example.chpl `

**Execution command:**
`./default-args-example -nl 1`

### Configuration Information

- Output of `chpl --version`: chpl version 1.24.1
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: none
CHPL_HWLOC: bundled
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`: gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0

Thanks Chapel Team!

### PR DESCRIPTION
**Pull request by bradcray**
This is a partial fix to problems reported in #7880, #17649, and #15164
in which the default arguments for a dynamically dispatched routine use
the defaults for the static class type's version of the method rather
than the dynamic class's.  The issue was that the methods that the
compiler introduces to represent the default values were tagged as
being `param` which prevented them from being dynamically dispatched.
This PR takes the approach of not applying the `param` tag to the
methods that compute the default values in order to have them dynamically
dispatch.  However, I call it a partial fix because if the argument itself
is a `param`, things still fall back to the old behavior.  Fixing this is a more
significant effort that I've spun off to issue #17683 (which also contains
some notes on how we might address it).

Big thanks to @mppf for providing the necessary context and pointers
to be able to solve this somewhat quickly.

Resolves #7880
Resolves #17649
Resolves #15164 (partially... the unresolved case remains in #17683)

### CODE REVIEW

**Comment by cassella** [comment]
You can probably also resolve #15164 with this.

**Comment by bradcray** [comment]
I meant to open this as a draft pull request because I wasn't confident it would pass testing, and it didn't (8 failures), so converting it to draft now.

Thanks Paul, I think you're right.

**Comment by bradcray** [comment]
Actually, I think I was being sloppy; I think it only solved the first of the two issues in #15164.

**Comment by bradcray** [comment]
@e-kayrakli : Interestingly, most of the errors I'm getting from this branch are from `--verify` compilations complaining about `_toFollower()`.  Would it be too optimistic for me to think "maybe these functions will just go away as Engin wraps up his 'remove tuples from zips' effort, and I should just wait until that's done before worrying more about them?

**Comment by e-kayrakli** [comment]
@bradcray -- that's a good question. Removing `_toFollower()` is not among the things that I want to accomplish in my current push.

In the current effort, I want to remove `_toFollowerZip()` and anything `fooZip` that peels off the tuple `(x,y,z)` to convert it to `(foo(x), foo(y), foo(z))` That being said, I have considered removing `toFollower` et al. as well. Though not deeply enough to venture a guess as to how easy it would be to do that.

**Comment by bradcray** [comment]
For my own reference as much as anything, here are the cases that are currently failling:

verification failures in toFollower
--------------------------------
[Error matching compiler output for classes/delete-free/lifetimes/tuple-vararg-in-lt-clause]
[Error matching compiler output for functions/iterators/vass/forall-in-standalone-iterator]
[Error matching compiler output for functions/promotion/racy-method-receiver]
[Error matching compiler output for types/records/diten/unusedExternRecordInProc]
[Error matching compiler output for types/records/init/nested-empty-record]
[Error matching .bad file for types/records/bharshbarger/recordArrCmp]

change in failed vectorization reporting
----------------------------------------
[Error matching program output for performance/vectorization/hintTests/vec-no-hint]

Next step is probably to compare what changed about these codes with my modest effort and see if I can tighten up the case in which it applies.

**Comment by bradcray** [comment]
@mppf: Would you be up for reviewing this one?

**Comment by bradcray** [comment]
Actually, maybe hold off on that.  Once I got `--verify` testing running the correct way, I see some failures.  Nightly `--verify` testing hasn't been doing the right thing lately, so there's a tiny chance that this isn't because of me, but I doubt it...

File: test/functions/default-arguments/default-argument-class-override3a.future [hunk 1]
```diff
@@ -0,0 +1,11 @@
+bug: dynamically dispatched methods with default param arguments wrong
+
+#15164
```
Comment (mppf) [inline test/functions/default-arguments/default-argument-class-override3a.future:3]: In the spirit of closing old issues and having more focused issues, I think we should close #15164 and create a new issue focused specifically on this case. You are welcome to copy my comment https://github.com/chapel-lang/chapel/issues/15164#issuecomment-834260625 into it.

Additionally I'd view adjusting the compiler to emit an error for this case (saying something like "not yet supported") as the immediate next step. Could you note that is the next step in the PR message?

**Review by mppf — APPROVED** [review]
Cool, thanks for fixing it!

File: test/functions/default-arguments/default-argument-class-override3a.future (hunk 1 above)
Comment (bradcray) [reply test/functions/default-arguments/default-argument-class-override3a.future:3]: Yeah, I had the same plan to close the bigger issue and making a more specific one.

I thought about adding the error message in this PR but (a) had a hard enough time landing this PR's fix without wanting to wrestle with it more and (b) am not convinced that we can actually generate such an error message without either breaking reasonable code or expending a fair amount of effort.  E.g., I think it should be reasonable to continue to use methods on classes that take `param` arguments that take `param` defaults (and believe we rely on such cases heavily), so we'd need to be sure to issue the error only in the case that there was dynamic dispatch involved.  Maybe that's not all that hard, but it seemed like more than a minor tweak to this wrapper code (e.g., I couldn't just check whether the formal we're creating the default for was `param` and issue the error if it is).

Comment (mppf) [reply test/functions/default-arguments/default-argument-class-override3a.future:3]: Yes, it's totally fine to do in a different PR. I suspect the override checking would be the place for it.

Comment (bradcray) [reply test/functions/default-arguments/default-argument-class-override3a.future:3]: I spawned off #17683 to capture the remaining part of this issue.  Just to make sure there are no misassumptions, I don't currently have any plans to work on it.

**Pull request merged**